### ISSUE TITLE
HA: Elector doesn't reconnect to etcd after EtcdClusterIdChanged

### ISSUE THREAD

**Issue by fasaxc** [Metaswitch/calico#556]
The EtcdClusterIdChanged exception is persistent.  Once hit, it will be raised on every subsequent watch(); we need to tear down the client and reconnect in that case (unless we change the behaviour in our fork of python-etcd but we've already made moves to upstream that change).

**Edit**: looks like we're going to address this in python-etcd rather than the Elector.

**Comment by Lukasa** [comment]
Well that's upsetting. =(

**Comment by Lukasa** [comment]
I'm not confident about doing this in the mechanism driver. It feels like the behaviour is wrong at the python-etcd level, and should be resolved there, rather than having this exception bust the etcd client forever.

**Comment by fasaxc** [comment]
We could raise the exception once and then update the stored cluster ID so we only raise it once per change?

**Comment by fasaxc** [comment]
I'll have to discuss with upstream though, I just updated the PR with them over the weekend :-(

**Comment by Lukasa** [comment]
@fasaxc That approach sounds better. =)

My intuition here is based on urllib3, which does its best to invalidate the minimal amount of state when an exception is thrown. Sometimes state does need to be invalidated, but usually it doesn't.

Note, btw, that right now none of the code in the plugin handles ClusterIDChanged. That may be a problem. =D

**Comment by fasaxc** [comment]
I mentioned the concerns about the fix in the upstream PR https://github.com/jplana/python-etcd/pull/92

**Comment by fasaxc** [comment]
PR open in the python-etcd repo to raise that exception only once.  That should help but I'm going to check the code to make sure we'll do the right thing.

**Comment by fasaxc** [comment]
Looks like the Elector is sound, it'll just restart from scratch if it hits EtcdClusterIdChanged.  Not sure about the mechanism driver.  I'll raise a separate issue.

**Comment by fasaxc** [comment]
Spoke too soon: what if the single cluster ID change exception goes to the plugin and the Elector never hears about it?  I think we need them to have separate clients rather than sharing a single client.

**Comment by Lukasa** [comment]
If the cluster ID change goes to the plugin the plugin needs to invalidate its etcd state. This *includes* beginning a new leader election. Not too complicated. =)
